src/Opdex.Platform.Application/EntryHandlers/Routers/GetSwapAmountInQueryHandler.cs

```diff
@@ -0,0 +1,71 @@
+
+            var tokenOutAmount = request.TokenOutAmount.ToSatoshis(tokenOut.Decimals);
+
+            var market = await _mediator.Send(new RetrieveMarketByAddressQuery(request.Market, findOrThrow: true), cancellationToken);
+            var router = await _mediator.Send(new RetrieveActiveMarketRouterByMarketIdQuery(market.Id, findOrThrow: true), cancellationToken);
+
+            UInt256 amountIn = request.IsSingleHopQuery ? await GetAmountInSingleHop(router, tokenIn, tokenOut, tokenOutAmount, cancellationToken)
+                                                        : await GetAmountInMultiHop(router, tokenIn, tokenOut, tokenOutAmount, cancellationToken);
+
+            return amountIn.ToDecimal(tokenIn.Decimals);
+        }
+
+        private async Task<UInt256> GetAmountInSingleHop(MarketRouter router, Token tokenIn, Token tokenOut, UInt256 tokenOutAmount, CancellationToken cancellationToken)
+        {
+            var tokenInIsCrs = tokenIn.Address == Address.Cirrus;
+
+            var srcTokenId = tokenInIsCrs ? tokenOut.Id : tokenIn.Id;
+            var liquidityPool = await _mediator.Send(new RetrieveLiquidityPoolBySrcTokenIdAndMarketIdQuery(srcTokenId, router.MarketId, findOrThrow: true), cancellationToken);
+
+            var reserves = await _mediator.Send(new CallCirrusGetOpdexLiquidityPoolReservesQuery(liquidityPool.Address), cancellationToken);
+            var reservesTokenIn = tokenInIsCrs ? reserves.Crs : reserves.Src;
+            var reservesTokenOut = tokenInIsCrs ? reserves.Src : reserves.Crs;
+
+            return await _mediator.Send(new CallCirrusGetAmountInStandardQuoteQuery(router.Address, tokenOutAmount, reservesTokenIn, reservesTokenOut), cancellationToken);
+        }
+
+        private async Task<UInt256> GetAmountInMultiHop(MarketRouter router, Token tokenIn, Token tokenOut, UInt256 tokenOutAmount, CancellationToken cancellationToken)
+        {
+            var liquidityPoolIn = await _mediator.Send(new RetrieveLiquidityPoolBySrcTokenIdAndMarketIdQuery(tokenIn.Id, router.MarketId, findOrThrow: true), cancellationToken);
+            var liquidityPoolOut = await _mediator.Send(new RetrieveLiquidityPoolBySrcTokenIdAndMarketIdQuery(tokenOut.Id, router.MarketId, findOrThrow: true), cancellationToken);
+
+            var reservesPoolIn = await _mediator.Send(new CallCirrusGetOpdexLiquidityPoolReservesQuery(liquidityPoolIn.Address), cancellationToken);
+            var reservesPoolOut = await _mediator.Send(new CallCirrusGetOpdexLiquidityPoolReservesQuery(liquidityPoolOut.Address), cancellationToken);
+
+            return await _mediator.Send(new CallCirrusGetAmountInMultiHopQuoteQuery(router.Address, tokenOutAmount,
+                                                                                    reservesPoolOut.Crs, reservesPoolOut.Src,
+                                                                                    reservesPoolIn.Crs, reservesPoolIn.Src), cancellationToken);
+        }
```

Comment:
Splitting into 4 queries with distinction between single/multi hop didn't quite seem right because the query is the same, with opposing validation on the token address. I've created a `RetrieveSwapAmountInQuery` and a `RetrieveSwapAmountOutQuery` and managed to add in more tests. Hopefully this is a happy medium.